### PR TITLE
ESP32-H2: add platform support for Thread end device

### PR DESCRIPTION
**Pull request by wqx6**
#### Problem
Add platform support for Thread end device on [ESP32-H2](https://www.espressif.com/en/news/ESP32_H2), the new BLE/15.4 combo SoC from Espressif.

#### Change overview
Add Thread stack in esp32 platform for our 15.4 chip ESP32-H2.

#### Testing
Test with ESP32-H2
1. BLE connection with python control.
2. Thread network provisioning

### CODE REVIEW

File: config/esp32/components/chip/CMakeLists.txt [hunk 1]
```diff
@@ -59,7 +63,7 @@ chip_gn_arg_append("esp32_cxx"             "\"${CMAKE_CXX_COMPILER}\"")
 chip_gn_arg_append("esp32_cpu"             "\"esp32\"")
 chip_gn_arg_bool("is_debug"                is_debug)
 
-if(CONFIG_BT_ENABLED)
+if(CONFIG_BT_NIMBLE_ENABLED OR CONFIG_BT_BLUEDROID_ENABLED)
```
Comment (chshu) [inline config/esp32/components/chip/CMakeLists.txt:66]: Is this change still necessary?

File: config/esp32/components/chip/Kconfig [hunk 2]
```diff
@@ -497,12 +497,12 @@ menu "CHIP Device Layer"
     endmenu
 
     menu "BLE Options"
-        visible if BT_ENABLED
+        visible if BT_NIMBLE_ENABLED || BT_BLUEDROID_ENABLED
```
Comment (chshu) [inline config/esp32/components/chip/Kconfig:500]: Please avoid the change here.

File: src/platform/ESP32/ConfigurationManagerImpl.cpp [hunk 3]
```diff
@@ -169,12 +174,16 @@ void ConfigurationManagerImpl::DoFactoryReset(intptr_t arg)
...
     // Restore WiFi persistent settings to default values.
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
     esp_err_t error = esp_wifi_restore();
     if (error != ESP_OK)
     {
         ChipLogError(DeviceLayer, "esp_wifi_restore() failed: %s", esp_err_to_name(error));
     }
-
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
```
Comment (chshu) [inline src/platform/ESP32/ConfigurationManagerImpl.cpp:184]: #elif CHIP_DEVICE_CONFIG_ENABLE_THREAD

File: src/platform/ESP32/ConnectivityManagerImpl.h [hunk 4]
```diff
@@ -178,18 +195,32 @@ inline bool ConnectivityManagerImpl::_CanStartWiFiScan()
...
+}
+#endif
+
 /**
  * Returns the public interface of the ConnectivityManager singleton object.
  *
  * Chip applications should use this to access features of the ConnectivityManager object
  * that are common to all platforms.
  */
+
```
Comment (chshu) [inline src/platform/ESP32/ConnectivityManagerImpl.h:223]: avoid unnecessary change.

File: src/platform/ESP32/ESPThreadConfig.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2018 Nest Labs, Inc.
```
Comment (chshu) [inline src/platform/ESP32/ESPThreadConfig.h:4]: remove this line.

**Review by chshu — COMMENTED** [review]
Please double verify that the change won't break any functionality of ESP32 and ESP32-C3 as Wi-Fi end devices.

File: config/esp32/components/chip/CMakeLists.txt (hunk 1 above)
Comment (wqx6) [reply config/esp32/components/chip/CMakeLists.txt]: Changed

File: config/esp32/components/chip/Kconfig (hunk 2 above)
Comment (wqx6) [reply config/esp32/components/chip/Kconfig]: Done

File: src/platform/ESP32/ConfigurationManagerImpl.cpp (hunk 3 above)
Comment (wqx6) [reply src/platform/ESP32/ConfigurationManagerImpl.cpp]: Done

File: src/platform/ESP32/ESPThreadConfig.h (hunk 5 above)
Comment (wqx6) [reply src/platform/ESP32/ESPThreadConfig.h]: Done

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/platform/ESP32/ConnectivityManagerImpl.h [hunk 6]
```diff
@@ -45,24 +53,32 @@ class PlatformManagerImpl;
...
+#endif
+                                      public Internal::GenericConnectivityManagerImpl<ConnectivityManagerImpl>
 {
 
     // Allow the ConnectivityManager interface class to delegate method calls to
     // the implementation methods provided by this class.
     friend class ConnectivityManager;
 
 private:
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
```
Comment (andy31415) [inline src/platform/ESP32/ConnectivityManagerImpl.h:79]: could we use some form of subclassing instead of `#ifdef` to compile-time switch code?

I would find an organization of `ConnectivityManagerImpl_Wifi.h`and `ConnectivitymanagerImpl_Thread.h` and cpp much easier to follow and review.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Please update the code to not `#ifdef` large amount of code. We had this in the system and inet layer and are suffering because of this.

File: config/esp32/components/chip/CMakeLists.txt [hunk 7]
```diff
@@ -29,7 +29,11 @@ if(NOT CHIP_ROOT)
     get_filename_component(CHIP_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../../.. REALPATH)
 endif()
 
-set(CHIP_REQURIE_COMPONENTS freertos lwip bt mdns mbedtls fatfs)
+set(CHIP_REQURIE_COMPONENTS freertos lwip bt nvs_flash mdns mbedtls fatfs)
```
Comment (dhrishi) [inline config/esp32/components/chip/CMakeLists.txt:32]: What has changed that previously `nvs_flash` was not required and now it is required?

File: src/platform/ESP32/CHIPDevicePlatformConfig.h [hunk 8]
```diff
@@ -41,6 +41,15 @@
...
 #define CHIP_DEVICE_CONFIG_DEFAULT_DEVICE_PRODUCT_REVISION CONFIG_DEFAULT_DEVICE_PRODUCT_REVISION
 #define CHIP_DEVICE_CONFIG_DEVICE_FIRMWARE_REVISION_STRING CONFIG_DEVICE_FIRMWARE_REVISION
+
+#if CONFIG_IDF_TARGET_ESP32H2
+#define CHIP_DEVICE_CONFIG_ENABLE_THREAD CONFIG_OPENTHREAD_ENABLED
+#define CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT CONFIG_OPENTHREAD_SRP_CLIENT
+#define CHIP_DEVICE_CONFIG_ENABLE_WIFI_AP 0
+#define CHIP_DEVICE_CONFIG_ENABLE_WIFI_STATION 0
+#define CHIP_DEVICE_CONFIG_ENABLE_WIFI_TELEMETRY 0
+#define CHIP_DEVICE_CONFIG_THREAD_ENABLE_CLI 0
```
Comment (dhrishi) [inline src/platform/ESP32/CHIPDevicePlatformConfig.h:51]: Should any of the above parameters be exposed through menuconfig? Or these values are never expected to change

File: src/platform/ESP32/ThreadStackManagerImpl.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+inline ThreadStackManager & ThreadStackMgr(void)
+{
+    return ThreadStackManagerImpl::sInstance;
+}
+
+/**
+ * Returns the platform-specific implementation of the ThreadStackManager singleton object.
+ *
+ * chip applications can use this to gain access to features of the ThreadStackManager
+ * that are specific to ESP32 platforms.
```
Comment (dhrishi) [inline src/platform/ESP32/ThreadStackManagerImpl.h:97]: Nit: ESP32 platform

File: src/platform/ESP32/ThreadStackManagerImpl.h [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+class ThreadStackManager;
+class ThreadStackManagerImpl;
+
+/**
+ * Concrate implementation of the ThreadStackManager singleton object for ESP32 platforms.
```
Comment (dhrishi) [inline src/platform/ESP32/ThreadStackManagerImpl.h:49]: Typo: Concrete, platform

File: src/platform/ESP32/ThreadStackManagerImpl.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides an implementation of the ThreadStackManager object for
+ *          ESP32 platforms.
```
Comment (dhrishi) [inline src/platform/ESP32/ThreadStackManagerImpl.cpp:21]: platform. Please check at all other places

**Comment by dhrishi** [comment]
@wqx6 Are there any changes required in the examples (all-clusters-app)? Documentation update?

File: src/platform/ESP32/ThreadStackManagerImpl.cpp (hunk 11 above)
Comment (wqx6) [reply src/platform/ESP32/ThreadStackManagerImpl.cpp]: Changed

File: src/platform/ESP32/ThreadStackManagerImpl.h (hunk 10 above)
Comment (wqx6) [reply src/platform/ESP32/ThreadStackManagerImpl.h]: Done

File: src/platform/ESP32/ThreadStackManagerImpl.h (hunk 9 above)
Comment (wqx6) [reply src/platform/ESP32/ThreadStackManagerImpl.h]: Done

File: src/platform/ESP32/CHIPDevicePlatformConfig.h (hunk 8 above)
Comment (wqx6) [reply src/platform/ESP32/CHIPDevicePlatformConfig.h]: Removed these definitions

File: config/esp32/components/chip/CMakeLists.txt (hunk 7 above)
Comment (wqx6) [reply config/esp32/components/chip/CMakeLists.txt]: Removed

**Comment by wqx6** [comment]
> @wqx6 Are there any changes required in the examples (all-clusters-app)? Documentation update?

the example changes for thread device will be added later.

**Review by chshu — APPROVED** [review]

File: src/platform/ESP32/ConnectivityManagerImpl.h (hunk 6 above)
Comment (wqx6) [reply src/platform/ESP32/ConnectivityManagerImpl.h]: The name of `ConnectivityManagerImpl.h` couldn't be changed because `CONNECTIVITYMANAGERIMPL_HEADER` is defined in [here](https://github.com/project-chip/connectedhomeip/blob/master/src/include/platform/ConnectivityManager.h#L267).  So I add  `ConnectivityManagerImpl_WiFi.h`, `ConnectivitymanagerImpl_Thread.h` and cpp files and modify `ConnectivityManagerImpl.h`:
```
#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
#include <platform/ESP32/ConnectivityManagerImpl_WiFi.h>
#elif CHIP_DEVICE_CONFIG_ENABLE_THREAD
#include <platform/ESP32/ConnectivityManagerImpl_Thread.h>
#endif
```

**Comment by wqx6** [comment]
PTAL @andy31415

**Review by dhrishi — APPROVED** [review]

**Review by andy31415 — COMMENTED** [review]
Could you ensure example apps (at least some) compile on the new platform?

./scripts/build/build_examples.py should work (so I can test compilation from vscode).

**Comment by chshu** [comment]
> Could you ensure example apps (at least some) compile on the new platform?
> 
> ./scripts/build/build_examples.py should work (so I can test compilation from vscode).

@andy31415 The ESP32-H2 IDF branch for Matter is not public yet, this PR mainly aims to enable Espressif's Thread platform support, the ESP32-H2 examples will be supported later when its IDF branch is public. 

By merging this PR, it helps to unblock our internal development process, that we can develop both Wi-Fi and Thread Matter devices software on the same code base.

**Comment by andy31415** [comment]
> > Could you ensure example apps (at least some) compile on the new platform?
> > ./scripts/build/build_examples.py should work (so I can test compilation from vscode).
> 
> @andy31415 The ESP32-H2 IDF branch for Matter is not public yet, this PR mainly aims to enable Espressif's Thread platform support, the ESP32-H2 examples will be supported later when its IDF branch is public.
> 
> By merging this PR, it helps to unblock our internal development process, that we can develop both Wi-Fi and Thread Matter devices software on the same code base.

My take on this is that we deny any platforms/versions that cannot be compiled by CHIP developers and CI (if we desire) without any need of special  configurations/separate downloads/signing NDAs or similar hurdles.

In this case, since this is an unreleased product, CHIP developers cannot compile it without having access to a private/restricted SDK build. This makes it unsuitable for mainline CHIP build. If you have access to some 'beta' SDK and can place that under the vscode (and/or esp32 version) we can reevaluate.

**Comment by woody-apple** [comment]
@andy31415 it looks like this compiles happily, but supports upcoming development.  

@wqx6 can you fix the merge conflicts here?

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
At some point we may want to make build_examples.py also support the new board type.
